node/include/cocaine/detail/service/node/engine.hpp

```diff
@@ -0,0 +1,166 @@
+#include "cocaine/service/node/slave/id.hpp"
+
+#include "cocaine/detail/service/node/rpc/slot.hpp"
+#include "cocaine/detail/service/node/slave.hpp"
+#include "cocaine/detail/service/node/slave/load.hpp"
+#include "cocaine/detail/service/node/stats.hpp"
+
+namespace cocaine {
+namespace detail {
+namespace service {
+namespace node {
+
+using cocaine::service::node::app::event_t;
+using cocaine::service::node::slave::id_t;
+
+using slave::control_t;
+using slave::load_t;
+
+class engine_t : public std::enable_shared_from_this<engine_t> {
+public:
+    enum class despawn_policy_t { graceful, force };
+
+    typedef std::deque<load_t> queue_type;
+    typedef std::unordered_map<std::string, slave_t> pool_type;
+
+    const std::unique_ptr<cocaine::logging::logger_t> log;
+
+    context_t& context;
+
+    /// Time point, when the overseer was created.
+    const std::chrono::system_clock::time_point birthstamp;
+
+    /// The application manifest.
+    const manifest_t manifest_;
+
+    /// The application profile.
+    synchronized<profile_t> profile_;
+
+    /// IO loop for timers and standard output fetchers.
+    std::shared_ptr<asio::io_service> loop;
+
+    /// Slave pool.
+    synchronized<pool_type> pool;
+    std::atomic<int> pool_target;
+
+    /// Pending queue.
+    synchronized<queue_type> queue;
+
+    /// Statistics.
+    stats_t stats;
+
+public:
+    engine_t(context_t& context, manifest_t manifest, profile_t profile,
+             std::shared_ptr<asio::io_service> loop);
+
+    ~engine_t();
+
+    /// Returns copy of the current manifest.
+    ///
+    /// Application's manifest is considered constant during all app's lifetime and can be
+    /// changed only through restarting.
+    auto manifest() const -> manifest_t;
+
+    /// Returns copy of the current profile, which is used to spawn new slaves.
+    ///
+    /// \note the current profile may change in any moment. Moveover some slaves can be in some kind
+    /// of transition state, i.e. migrating from one profile to another.
+    auto profile() const -> profile_t;
+
+    /// Returns application total uptime in seconds.
+    auto uptime() const -> std::chrono::seconds;
+
+    /// Returns the complete info about how the application works using json-like object.
+    auto info(io::node::info::flags_t flags) const -> dynamic_t::object_t;
+
+    // Modifiers.
+
+    /// Enqueues the new event into the most appropriate slave.
+    ///
+    /// The event will be put into the queue if there are no slaves available at this moment or all
+    /// of them are busy.
+    ///
+    /// \return the dispatch object, which is ready for processing the appropriate protocol
+    /// messages.
+    ///
+    /// \param downstream represents the [Client <- Worker] stream.
+    /// \param event an invocation event.
+    /// \param id represents slave id to be enqueued (may be none, which means any slave).
+    std::shared_ptr<client_rpc_dispatch_t> enqueue(
```

Comment:
Only application I can find is using decltype of parameter as return type, but ok, let's use this cyber cool, modern C++!